### ISSUE TITLE
Webpage.uploadFile not working in phantomjs 2.0

### ISSUE THREAD

**Issue by loduis** [ariya/phantomjs#12506]
This is in linux (Ubuntu)

```javascript

var
  page     = require('webpage').create(),
  filename = '/media/loduis/DATOS/projects/ghostdriver/test/javascript/local.jpg';



page.onFilePicker = function(oldFile) {

  console.log('Listen event.');

  return filename;
};


page.open('data:text/html,' +
  '<body><form method="POST" enctype="multipart/form-data">' +
  'File to upload: <input type="file" name="upfile" style="width: 200px;height: 100px; background: gray">' +
  '</form></body>', function(status) {

  console.log(status);
  page.uploadFile('input[name=upfile]', filename);

  page.render('test.png');
});
```
The result on phantomjs 1.9.7 is 
![test2](https://cloud.githubusercontent.com/assets/451862/4051160/1b9d2e1a-2d5f-11e4-83c0-63720d82ad09.png)

while 2.0 is
![test](https://cloud.githubusercontent.com/assets/451862/4051121/bd36a7d4-2d5e-11e4-9451-693f5dfa91a9.png)

another important thing is that it is never fired onFilePicker

**Comment by SpenserJ** [comment]
This has solved the issue for me as well. Thank you @burkestar :+1:

**Comment by mhoran** [comment]
It would be great to get a fix for this rolled into upstream Qt. We had originally filed a bug in WebKit here: https://bugs.webkit.org/show_bug.cgi?id=75385. It was closed when all Qt related bugs were bulk closed in the WebKit bug tracker.

In capybara-webkit, we work around this by using native clicks to submit file uploads.

**Comment by xaviershay** [comment]
fwiw, poltergeist spec showing the error if anyone was looking for a cheap way to repro: https://github.com/teampoltergeist/poltergeist/pull/585

**Comment by dmitryrechkin** [comment]
I've tried javascript code example from this issue in phantomjs 2 which is published your website and upload file is not working. Was it fixed at all?

**Comment by Nek-** [comment]
I don't really know what change the patch, but it worked for me ! I can now upload via webdriver !

Thank you @burkestar

**Comment by vassilevsky** [comment]
Works for me too. Thanks @burkestar!

**Comment by xaviershay** [comment]
Just updating that this isn't fixed in the latest version installed by `brew`.

Seems like we need to patch QT upstream? https://bugs.webkit.org/show_bug.cgi?id=75385 doesn't necessarily seem appropriate, since it was file in 2012 and this issue only just surfaced?

**Comment by stevepm** [comment]
Have there been any updates to fix this besides the patch?

**Comment by leeight** [comment]
:+1:

**Comment by leeight** [comment]
:+1:

**Comment by stevepm** [comment]
I forked and made the update and it works great for me if anyone wants to build the binary off of my fork: https://github.com/stevepm/phantomjs/tree/2.0

**Comment by dmitryrechkin** [comment]
Is it possible to merge your fix to the main repository? Which commit fixes this issue? Thanks

**Comment by Vitallium** [comment]
Yes, 2.0.1 version will receive this fix

**Comment by dmitryrechkin** [comment]
Do you know when approximately 2.0.1 will be published?

**Comment by stevepm** [comment]
@dmitryrechkin I only made one commit. See here: https://github.com/stevepm/phantomjs/commit/5f663b8b39138c0510a7acac153a86492373988f

**Comment by jondkinney** [comment]
@stevepm thanks for pointing us at your fork. I'm going to try to compile a binary from your fork for OS X Yosemite. Unless anyone else already done this?

**Comment by RamyaBNayak** [comment]
Could anyone please share the compiled exe for windows for this fix? I am having difficulty compiling.

**Comment by chipmunkgeek** [comment]
:+1: on the suggested patch.  Thank you!

**Comment by leopucci** [comment]
Hi @stevepm , do you have a windows binary of your fork? (https://github.com/stevepm/phantomjs/tree/2.0) Could you share it?
Thanks

**Comment by skakri** [comment]
I have Ubuntu/Arch/OSX binaries @ https://github.com/skakri/phantomjs/releases/tag/2.0.1-regression-12506 , if someone needs those.

**Comment by leopucci** [comment]
Thanks @skakri, your build have the patch for upload file gesture :+1: i´ll try to build on the weekend.

**Comment by skakri** [comment]
@leopucci if you manage to build it on Windows x68_64, feel free to send me binary link @ kristaps.karlsons@gmail.com. I'll add it to downloads.

**Comment by leopucci** [comment]
Ok! :) :8ball:

**Comment by astrosuf** [comment]
Hey,

has a windows patch for the upload file bug been created? Could someone direct me to the patch this will be extremely helpful!

**Comment by cantonic** [comment]
I compiled a binary for ubuntu 12.04 64bit using the source from @skakri on https://github.com/skakri/phantomjs/releases/tag/2.0.1-regression-12506
The ubuntu binary in the repo threw an error when using it on TravisCI: `phantomjs: error while loading shared libraries: libicui18n.so.52: cannot open shared object file: No such file or directory`.

I have uploaded the binary on https://s3-eu-west-1.amazonaws.com/adveli-public/phantomjs-2.0.1-regression-12506-ubuntu-12.04_64.tar.bz2

**Comment by leopucci** [comment]
Sorry i did not had time to compile for windows yet as i told.
Em 17/07/2015 00:19, "Can Edremitoglu" <notifications@github.com> escreveu:

**Comment by samir-mahendra** [comment]
@skakri Thanks for sharing the source.  My colleague compiled it for windows, binary is on Dropbox here: https://www.dropbox.com/s/u71vipap8bnm72u/windows-x86_64-phantomjs.exe?dl=0

**Comment by skakri** [comment]
@samir-mahendra – thanks, added. @cantonic sorry, I've compiled it for Ubuntu 14.04.

**Comment by astrosuf** [comment]
@samir-mahendra Thanks!!

**Comment by geminiyellow** [comment]
still not working.

**Comment by samir-mahendra** [comment]
@geminiyellow Can you give more details on which OS you are using, and the specific test that fails for you?  The test script I used to test if it's working is below.  You have to run it as follows:

phantomjs --ssl-protocol=any testscript.js <testfile>

You can then check if the file got uploaded by going to: https://encodable.com/uploaddemo/?action=listfiles

    // testscript.js
    var webpage = require('webpage'),
    system = require('system'),
    fname;

    if (system.args.length !== 2) {
        console.log('Usage: uploadtest.js filename');
        phantom.exit(1);
    } else {
        fname = system.args[1];
        var page = webpage.create();
    
        page.onConsoleMessage = function(msg) {
            console.log(msg);
        };
    
        page.open("https://encodable.com/uploaddemo/", function (status) {
            console.log("Page opened, ready to upload file.  Status is", status);
            page.uploadFile('input[name=uploadname1]', fname);
            page.evaluate(function () {
                document.forms[0].submit();
            });
            window.setTimeout(function () {
                phantom.exit();
            }, 3000);
        });
    }

**Comment by christopherbalz** [comment]
+1

**Comment by dangerous** [comment]
+1

**Comment by skakri** [comment]
Please stop spamming "+1", it adds clutter.

**Comment by Vitallium** [comment]
We have a proposal fix for this issue. And it will be landed in 2.0.1.

**Comment by Vitallium** [comment]
Fixed in master.
